File: leetcode/1143.go
func longestCommonSubsequence(text1 string, text2 string) int {
	// if len(text1) < len(text2) {
	//     text1, text2 = text2, text1
	// }
	m, n := len(text1), len(text2)
	ans := make([][]int, 2)
	for i := 0; i < 2; i++ {
		ans[i] = make([]int, n+1)
	}
	for i, ch1 := range text1 {
		for j, ch2 := range text2 {
			if ch1 == ch2 {
				ans[(i+1)&1][j+1] = ans[i&1][j] + 1
			} else if ans[i&1][j+1] < ans[(i+1)&1][j] {
				ans[(i+1)&1][j+1] = ans[(i+1)&1][j]
			} else {
				ans[(i+1)&1][j+1] = ans[i&1][j+1]
			}
		}
	}
	return ans[m&1][n]
}

File: leetcode/1143.cpp
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        int m = text1.size(),n = text2.size();
        int dp[2][n+1];
        fill(dp[0],dp[0]+n+1,0);
        fill(dp[1],dp[1]+n+1,0);
        for (int i = 0; i < m;i++){
            for (int j = 0; j < n;j++){
                if (text1[i] == text2[j]) {
                    dp[(i+1)&1][j+1] = dp[i&1][j] + 1;
                } else {
                    dp[(i+1)&1][j+1] = max(dp[(i+1)&1][j],dp[i&1][j+1]);
                }
            }
        }
        return dp[m&1][n];
    }
};